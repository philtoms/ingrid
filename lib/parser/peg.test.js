const bindings = require('../test/bindings');
const identity = require('../utils/identity');
const baseActions = require('./actions');
const rules = require('./rules');
const grammar = require('./grammar');
const peg = require('./peg');

const log = (res) => console.log(JSON.stringify(res, null, 2)) || res;

describe('parser', () => {
  let parse;
  beforeEach(() => {
    frame = { bindings };
    parse = peg()(grammar(rules), baseActions(identity));
  });

  it('selector', () => {
    expect(parse('a')).toMatchSnapshot();
  });
  it('selector aggregation', () => {
    expect(parse('a, b, c')).toMatchSnapshot();
  });
  it('full selector', () => {
    expect(parse('x:a(b)?, c')).toMatchSnapshot();
  });
  it('selector quantifier', () => {
    expect(parse('a*')).toMatchSnapshot();
  });
  it('quantifier', () => {
    expect(parse('A?')).toMatchSnapshot();
  });
  it('limit quantifier', () => {
    expect(parse('A{1}')).toMatchSnapshot();
  });
  it('range quantifier', () => {
    expect(parse('A{0,2}')).toMatchSnapshot();
  });
  it('alias str', () => {
    expect(parse(`'a b':x`)).toMatchSnapshot();
  });
  it('str - escaped', () => {
    expect(parse("'\\'b'")).toMatchSnapshot();
  });
  it('message', () => {
    expect(parse('A(x)')).toMatchSnapshot();
  });
  it('empty message', () => {
    expect(parse('A()')).toMatchSnapshot();
  });
  it('quantified message', () => {
    expect(parse('A(x+)')).toMatchSnapshot();
  });
  it('quantified annon message', () => {
    expect(parse('A(*)')).toMatchSnapshot();
  });
  it('aliased message', () => {
    expect(parse('A(X:1)')).toMatchSnapshot();
  });
  it('aliased annon message', () => {
    expect(parse('A(id:*)')).toMatchSnapshot();
  });
  it('aliased empty message', () => {
    expect(parse('A(id:)')).toMatchSnapshot();
  });
  it('message intention', () => {
    expect(parse('A is postfix(1, 2)')).toMatchSnapshot();
  });
  it('block', () => {
    expect(parse('A {B}')).toMatchSnapshot();
  });
  it('block aggregate', () => {
    expect(parse('A {B} {C}')).toMatchSnapshot();
  });
  it('block aggregate interleaved', () => {
    expect(parse('A {B} is B {C} is C')).toMatchSnapshot();
  });
  it('statement blocks', () => {
    expect(parse('A B {C} D {E}')).toMatchSnapshot();
  });
  it('regex selector', () => {
    expect(parse('/123/')).toMatchSnapshot();
  });
  it('regex selector ref', () => {
    expect(parse('A is /123/')).toMatchSnapshot();
  });
  it('multi-line', () => {
    const nested = `
    A1 is
    A2
`;
    expect(parse(nested)).toMatchSnapshot();
  });
  it('nested', () => {
    const nested = `
      A1 {
        B1 is 123 {
          C1 is D1 {
            D1 is 456
          }
        }
      }
      `;
    expect(parse(nested)).toMatchSnapshot();
  });
});
