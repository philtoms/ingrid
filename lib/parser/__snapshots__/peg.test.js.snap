// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parser alias str 1`] = `
Array [
  Object {
    "intentions": Array [],
    "selectors": Array [
      Object {
        "message": null,
        "token": Object {
          "alias": "a b",
          "ident": "x",
          "quantifier": null,
          "rule": "token",
          "type": "tag",
        },
      },
    ],
  },
]
`;

exports[`parser aliased annon message 1`] = `
Array [
  Object {
    "intentions": Array [],
    "selectors": Array [
      Object {
        "message": Object {
          "selector": Array [
            Object {
              "token": Object {
                "alias": "id",
                "quantifier": "*",
                "type": Symbol(empty),
              },
            },
          ],
        },
        "token": Object {
          "ident": "A",
          "quantifier": null,
          "rule": "token",
          "type": "tag",
        },
      },
    ],
  },
]
`;

exports[`parser aliased empty message 1`] = `
Array [
  Object {
    "intentions": Array [],
    "selectors": Array [
      Object {
        "message": Object {
          "selector": Array [
            Object {
              "token": Object {
                "alias": "id",
                "type": Symbol(empty),
              },
            },
          ],
        },
        "token": Object {
          "ident": "A",
          "quantifier": null,
          "rule": "token",
          "type": "tag",
        },
      },
    ],
  },
]
`;

exports[`parser aliased message 1`] = `
Array [
  Object {
    "intentions": Array [],
    "selectors": Array [
      Object {
        "message": Object {
          "selector": Array [
            Object {
              "message": null,
              "token": Object {
                "alias": "X",
                "ident": "1",
                "quantifier": null,
                "rule": "token",
                "type": "tag",
              },
            },
          ],
        },
        "token": Object {
          "ident": "A",
          "quantifier": null,
          "rule": "token",
          "type": "tag",
        },
      },
    ],
  },
]
`;

exports[`parser block 1`] = `
Array [
  Object {
    "intentions": Array [
      Object {
        "intent": "block",
        "selector": Array [
          Object {
            "intentions": Array [],
            "selectors": Array [
              Object {
                "message": null,
                "token": Object {
                  "ident": "B",
                  "quantifier": null,
                  "rule": "token",
                  "type": "tag",
                },
              },
            ],
          },
        ],
      },
    ],
    "selectors": Array [
      Object {
        "message": null,
        "token": Object {
          "ident": "A",
          "quantifier": null,
          "rule": "token",
          "type": "tag",
        },
      },
    ],
  },
]
`;

exports[`parser block aggregate 1`] = `
Array [
  Object {
    "intentions": Array [
      Object {
        "intent": "block",
        "selector": Array [
          Object {
            "intentions": Array [],
            "selectors": Array [
              Object {
                "message": null,
                "token": Object {
                  "ident": "B",
                  "quantifier": null,
                  "rule": "token",
                  "type": "tag",
                },
              },
            ],
          },
        ],
      },
      Object {
        "intent": "block",
        "selector": Array [
          Object {
            "intentions": Array [],
            "selectors": Array [
              Object {
                "message": null,
                "token": Object {
                  "ident": "C",
                  "quantifier": null,
                  "rule": "token",
                  "type": "tag",
                },
              },
            ],
          },
        ],
      },
    ],
    "selectors": Array [
      Object {
        "message": null,
        "token": Object {
          "ident": "A",
          "quantifier": null,
          "rule": "token",
          "type": "tag",
        },
      },
    ],
  },
]
`;

exports[`parser block aggregate interleaved 1`] = `
Array [
  Object {
    "intentions": Array [
      Object {
        "intent": "block",
        "selector": Array [
          Object {
            "intentions": Array [],
            "selectors": Array [
              Object {
                "message": null,
                "token": Object {
                  "ident": "B",
                  "quantifier": null,
                  "rule": "token",
                  "type": "tag",
                },
              },
            ],
          },
        ],
      },
      Object {
        "intent": "is",
        "selector": Array [
          Object {
            "message": null,
            "token": Object {
              "ident": "B",
              "quantifier": null,
              "rule": "token",
              "type": "tag",
            },
          },
        ],
      },
      Object {
        "intent": "block",
        "selector": Array [
          Object {
            "intentions": Array [],
            "selectors": Array [
              Object {
                "message": null,
                "token": Object {
                  "ident": "C",
                  "quantifier": null,
                  "rule": "token",
                  "type": "tag",
                },
              },
            ],
          },
        ],
      },
      Object {
        "intent": "is",
        "selector": Array [
          Object {
            "message": null,
            "token": Object {
              "ident": "C",
              "quantifier": null,
              "rule": "token",
              "type": "tag",
            },
          },
        ],
      },
    ],
    "selectors": Array [
      Object {
        "message": null,
        "token": Object {
          "ident": "A",
          "quantifier": null,
          "rule": "token",
          "type": "tag",
        },
      },
    ],
  },
]
`;

exports[`parser empty message 1`] = `
Array [
  Object {
    "intentions": Array [],
    "selectors": Array [
      Object {
        "message": Object {
          "selector": Array [
            Object {
              "token": Object {
                "type": Symbol(empty),
              },
            },
          ],
        },
        "token": Object {
          "ident": "A",
          "quantifier": null,
          "rule": "token",
          "type": "tag",
        },
      },
    ],
  },
]
`;

exports[`parser full selector 1`] = `
Array [
  Object {
    "intentions": Array [],
    "selectors": Array [
      Object {
        "message": Object {
          "selector": Array [
            Object {
              "message": null,
              "token": Object {
                "ident": "b",
                "quantifier": null,
                "rule": "token",
                "type": "tag",
              },
            },
          ],
        },
        "token": Object {
          "alias": "x",
          "ident": "a",
          "quantifier": "?",
          "rule": "token",
          "type": "tag",
        },
      },
      Object {
        "message": null,
        "token": Object {
          "ident": "c",
          "quantifier": null,
          "rule": "token",
          "type": "tag",
        },
      },
    ],
  },
]
`;

exports[`parser limit quantifier 1`] = `
Array [
  Object {
    "intentions": Array [],
    "selectors": Array [
      Object {
        "message": null,
        "token": Object {
          "ident": "A",
          "quantifier": Object {
            "max": "1",
            "min": "1",
          },
          "rule": "token",
          "type": "tag",
        },
      },
    ],
  },
]
`;

exports[`parser message 1`] = `
Array [
  Object {
    "intentions": Array [],
    "selectors": Array [
      Object {
        "message": Object {
          "selector": Array [
            Object {
              "message": null,
              "token": Object {
                "ident": "x",
                "quantifier": null,
                "rule": "token",
                "type": "tag",
              },
            },
          ],
        },
        "token": Object {
          "ident": "A",
          "quantifier": null,
          "rule": "token",
          "type": "tag",
        },
      },
    ],
  },
]
`;

exports[`parser message intention 1`] = `
Array [
  Object {
    "intentions": Array [
      Object {
        "intent": "is",
        "selector": Array [
          Object {
            "message": Object {
              "selector": Array [
                Object {
                  "message": null,
                  "token": Object {
                    "ident": "1",
                    "quantifier": null,
                    "rule": "token",
                    "type": "tag",
                  },
                },
                Object {
                  "message": null,
                  "token": Object {
                    "ident": "2",
                    "quantifier": null,
                    "rule": "token",
                    "type": "tag",
                  },
                },
              ],
            },
            "token": Object {
              "ident": "postfix",
              "quantifier": null,
              "rule": "token",
              "type": "tag",
            },
          },
        ],
      },
    ],
    "selectors": Array [
      Object {
        "message": null,
        "token": Object {
          "ident": "A",
          "quantifier": null,
          "rule": "token",
          "type": "tag",
        },
      },
    ],
  },
]
`;

exports[`parser multi-line 1`] = `
Array [
  Object {
    "intentions": Array [
      Object {
        "intent": "is",
        "selector": Array [
          Object {
            "message": null,
            "token": Object {
              "ident": "A2",
              "quantifier": null,
              "rule": "token",
              "type": "tag",
            },
          },
        ],
      },
    ],
    "selectors": Array [
      Object {
        "message": null,
        "token": Object {
          "ident": "A1",
          "quantifier": null,
          "rule": "token",
          "type": "tag",
        },
      },
    ],
  },
]
`;

exports[`parser nested 1`] = `
Array [
  Object {
    "intentions": Array [
      Object {
        "intent": "block",
        "selector": Array [
          Object {
            "intentions": Array [
              Object {
                "intent": "is",
                "selector": Array [
                  Object {
                    "message": null,
                    "token": Object {
                      "ident": "123",
                      "quantifier": null,
                      "rule": "token",
                      "type": "tag",
                    },
                  },
                ],
              },
              Object {
                "intent": "block",
                "selector": Array [
                  Object {
                    "intentions": Array [
                      Object {
                        "intent": "is",
                        "selector": Array [
                          Object {
                            "message": null,
                            "token": Object {
                              "ident": "D1",
                              "quantifier": null,
                              "rule": "token",
                              "type": "tag",
                            },
                          },
                        ],
                      },
                      Object {
                        "intent": "block",
                        "selector": Array [
                          Object {
                            "intentions": Array [
                              Object {
                                "intent": "is",
                                "selector": Array [
                                  Object {
                                    "message": null,
                                    "token": Object {
                                      "ident": "456",
                                      "quantifier": null,
                                      "rule": "token",
                                      "type": "tag",
                                    },
                                  },
                                ],
                              },
                            ],
                            "selectors": Array [
                              Object {
                                "message": null,
                                "token": Object {
                                  "ident": "D1",
                                  "quantifier": null,
                                  "rule": "token",
                                  "type": "tag",
                                },
                              },
                            ],
                          },
                        ],
                      },
                    ],
                    "selectors": Array [
                      Object {
                        "message": null,
                        "token": Object {
                          "ident": "C1",
                          "quantifier": null,
                          "rule": "token",
                          "type": "tag",
                        },
                      },
                    ],
                  },
                ],
              },
            ],
            "selectors": Array [
              Object {
                "message": null,
                "token": Object {
                  "ident": "B1",
                  "quantifier": null,
                  "rule": "token",
                  "type": "tag",
                },
              },
            ],
          },
        ],
      },
    ],
    "selectors": Array [
      Object {
        "message": null,
        "token": Object {
          "ident": "A1",
          "quantifier": null,
          "rule": "token",
          "type": "tag",
        },
      },
    ],
  },
]
`;

exports[`parser quantified annon message 1`] = `
Array [
  Object {
    "intentions": Array [],
    "selectors": Array [
      Object {
        "message": Object {
          "selector": Array [
            Object {
              "token": Object {
                "quantifier": "*",
                "type": Symbol(empty),
              },
            },
          ],
        },
        "token": Object {
          "ident": "A",
          "quantifier": null,
          "rule": "token",
          "type": "tag",
        },
      },
    ],
  },
]
`;

exports[`parser quantified message 1`] = `
Array [
  Object {
    "intentions": Array [],
    "selectors": Array [
      Object {
        "message": Object {
          "selector": Array [
            Object {
              "message": null,
              "token": Object {
                "ident": "x",
                "quantifier": "+",
                "rule": "token",
                "type": "tag",
              },
            },
          ],
        },
        "token": Object {
          "ident": "A",
          "quantifier": null,
          "rule": "token",
          "type": "tag",
        },
      },
    ],
  },
]
`;

exports[`parser quantifier 1`] = `
Array [
  Object {
    "intentions": Array [],
    "selectors": Array [
      Object {
        "message": null,
        "token": Object {
          "ident": "A",
          "quantifier": "?",
          "rule": "token",
          "type": "tag",
        },
      },
    ],
  },
]
`;

exports[`parser range quantifier 1`] = `
Array [
  Object {
    "intentions": Array [],
    "selectors": Array [
      Object {
        "message": null,
        "token": Object {
          "ident": "A",
          "quantifier": Object {
            "max": "2",
            "min": "0",
          },
          "rule": "token",
          "type": "tag",
        },
      },
    ],
  },
]
`;

exports[`parser regex selector 1`] = `
Array [
  Object {
    "intentions": Array [],
    "selectors": Array [
      Object {
        "message": null,
        "token": Object {
          "ident": "/123",
          "quantifier": null,
          "rule": "token",
          "type": "regex",
        },
      },
    ],
  },
]
`;

exports[`parser regex selector ref 1`] = `
Array [
  Object {
    "intentions": Array [
      Object {
        "intent": "is",
        "selector": Array [
          Object {
            "message": null,
            "token": Object {
              "ident": "/123",
              "quantifier": null,
              "rule": "token",
              "type": "regex",
            },
          },
        ],
      },
    ],
    "selectors": Array [
      Object {
        "message": null,
        "token": Object {
          "ident": "A",
          "quantifier": null,
          "rule": "token",
          "type": "tag",
        },
      },
    ],
  },
]
`;

exports[`parser selector 1`] = `
Array [
  Object {
    "intentions": Array [],
    "selectors": Array [
      Object {
        "message": null,
        "token": Object {
          "ident": "a",
          "quantifier": null,
          "rule": "token",
          "type": "tag",
        },
      },
    ],
  },
]
`;

exports[`parser selector aggregation 1`] = `
Array [
  Object {
    "intentions": Array [],
    "selectors": Array [
      Object {
        "message": null,
        "token": Object {
          "ident": "a",
          "quantifier": null,
          "rule": "token",
          "type": "tag",
        },
      },
      Object {
        "message": null,
        "token": Object {
          "ident": "b",
          "quantifier": null,
          "rule": "token",
          "type": "tag",
        },
      },
      Object {
        "message": null,
        "token": Object {
          "ident": "c",
          "quantifier": null,
          "rule": "token",
          "type": "tag",
        },
      },
    ],
  },
]
`;

exports[`parser selector quantifier 1`] = `
Array [
  Object {
    "intentions": Array [],
    "selectors": Array [
      Object {
        "message": null,
        "token": Object {
          "ident": "a",
          "quantifier": "*",
          "rule": "token",
          "type": "tag",
        },
      },
    ],
  },
]
`;

exports[`parser statement blocks 1`] = `
Array [
  Object {
    "intentions": Array [],
    "selectors": Array [
      Object {
        "message": null,
        "token": Object {
          "ident": "A",
          "quantifier": null,
          "rule": "token",
          "type": "tag",
        },
      },
    ],
  },
  Object {
    "intentions": Array [
      Object {
        "intent": "block",
        "selector": Array [
          Object {
            "intentions": Array [],
            "selectors": Array [
              Object {
                "message": null,
                "token": Object {
                  "ident": "C",
                  "quantifier": null,
                  "rule": "token",
                  "type": "tag",
                },
              },
            ],
          },
        ],
      },
    ],
    "selectors": Array [
      Object {
        "message": null,
        "token": Object {
          "ident": "B",
          "quantifier": null,
          "rule": "token",
          "type": "tag",
        },
      },
    ],
  },
  Object {
    "intentions": Array [
      Object {
        "intent": "block",
        "selector": Array [
          Object {
            "intentions": Array [],
            "selectors": Array [
              Object {
                "message": null,
                "token": Object {
                  "ident": "E",
                  "quantifier": null,
                  "rule": "token",
                  "type": "tag",
                },
              },
            ],
          },
        ],
      },
    ],
    "selectors": Array [
      Object {
        "message": null,
        "token": Object {
          "ident": "D",
          "quantifier": null,
          "rule": "token",
          "type": "tag",
        },
      },
    ],
  },
]
`;

exports[`parser str - escaped 1`] = `
Array [
  Object {
    "intentions": Array [],
    "selectors": Array [
      Object {
        "message": null,
        "token": Object {
          "ident": "\\\\'b",
          "quantifier": null,
          "rule": "token",
          "type": "tag",
        },
      },
    ],
  },
]
`;
